### PR TITLE
drivers: Add Atmel SAM family I2C (TWI) driver

### PR DESCRIPTION
**Pull request by mnkp**
Added I2C bus (TWI) driver for Atmel SAM MCU family. Only
I2C Master Mode with 7 bit addressing is currently supported.

Origin: Original

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

### CODE REVIEW

**Comment by bryan-hunt** [comment]
Pinmap/pinmux should be moved to the board rather than reside under the SOC.

**Comment by mnkp** [comment]
@bryan-hunt the pinmux code configures pins as defined by the Atmel SAM SoC family datasheets. It has nothing to do with the specifics of the board implementation/routing. Pins are configured only when the device instance is initialized, otherwise they are not touched.

**Comment by bryan-hunt** [comment]
Yes it does match the SAM3/SAM4S series datasheets where the pinmux is many to one but for the SAMD or SAME series with more conventional pinmuxing the map is many to many so I was recommending the change for convention with the number of other boards already in the zephyr system (e.g. stm32f429i_disc1, arduino_101, frdm_k64f, etc). As I'm working on both the SAM4s Xplained Pro and SAMD21 Xplained Pro ports I'll leave the pinmap where it is in my pull request for the SAM4S and leave the question of convention up to the ARM arch guys.

**Comment by mnkp** [comment]
Good point. If a signal can be routed to more than one pin on the SoC the choice should be made by the board config. This is handled currently by selecting a Kconfig option. E.g. to choose UART1 TXD pin on SAME70 SoC one has to set one of UART_SAM_PORT_1_PIN_TX_PA4, UART_SAM_PORT_1_PIN_TX_PA6, UART_SAM_PORT_1_PIN_TX_PD26 Kconfig options. This has to be ported to DTS. It's still on the TODO list.

By keeping the actual pin muxing code local to SoC we have only one copy of it.

File: drivers/i2c/i2c_sam_twi.c [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+	if (config & I2C_ADDR_10_BITS) {
+		SYS_LOG_ERR("I2C 10-bit addressing is currently not supported");
+		SYS_LOG_ERR("Please submit a patch");
+		return -EIO;
+	}
+
+	cfg.raw = config;
+
+	/* Configure clock */
+	switch (cfg.bits.speed) {
```
Comment (galak) [inline drivers/i2c/i2c_sam_twi.c]: can we avoid using cfg.bits.speed and instead doing something like:

speed = (config & I2C_SPEED_MASK) >> I2C_SPEED_SHIFT;
switch (speed)

File: drivers/i2c/i2c_sam_twi.c [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+};
+
+/* I2C0 */
+
+#ifdef CONFIG_I2C_0
+static struct device DEVICE_NAME_GET(i2c0_sam);
+
+static void i2c0_sam_irq_config(void)
+{
+	IRQ_CONNECT(TWI0_IRQn, CONFIG_I2C_0_IRQ_PRI, i2c_sam_twi_isr,
```
Comment (galak) [inline drivers/i2c/i2c_sam_twi.c]: we should get IRQ data from DTS

File: drivers/i2c/i2c_sam_twi.c [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+		    DEVICE_GET(i2c0_sam), 0);
+}
+
+static const struct soc_gpio_pin pins_twi0[] = PINS_TWI0;
+
+static const struct i2c_sam_twi_dev_cfg i2c0_sam_config = {
+	.regs = TWI0,
```
Comment (galak) [inline drivers/i2c/i2c_sam_twi.c]: we should get register address from DTS

File: drivers/i2c/i2c_sam_twi.c [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+#endif
+
+/* I2C1 */
+
+#ifdef CONFIG_I2C_1
+static struct device DEVICE_NAME_GET(i2c1_sam);
+
+static void i2c1_sam_irq_config(void)
+{
+	IRQ_CONNECT(TWI1_IRQn, CONFIG_I2C_1_IRQ_PRI, i2c_sam_twi_isr,
```
Comment (galak) [inline drivers/i2c/i2c_sam_twi.c]: use DTS for IRQ info

File: drivers/i2c/i2c_sam_twi.c [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+		    DEVICE_GET(i2c1_sam), 0);
+}
+
+static const struct soc_gpio_pin pins_twi1[] = PINS_TWI1;
+
+static const struct i2c_sam_twi_dev_cfg i2c1_sam_config = {
+	.regs = TWI1,
```
Comment (galak) [inline drivers/i2c/i2c_sam_twi.c]: use DTS for reg addr.

File: drivers/i2c/i2c_sam_twi.c [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+		    DEVICE_GET(i2c0_sam), 0);
+}
+
+static const struct soc_gpio_pin pins_twi0[] = PINS_TWI0;
+
+static const struct i2c_sam_twi_dev_cfg i2c0_sam_config = {
+	.regs = TWI0,
+	.irq_config = i2c0_sam_irq_config,
+	.periph_id = ID_TWI0,
+	.irq_id = TWI0_IRQn,
```
Comment (galak) [inline drivers/i2c/i2c_sam_twi.c]: use DTS for IRQ info

File: drivers/i2c/i2c_sam_twi.c [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+		    DEVICE_GET(i2c1_sam), 0);
+}
+
+static const struct soc_gpio_pin pins_twi1[] = PINS_TWI1;
+
+static const struct i2c_sam_twi_dev_cfg i2c1_sam_config = {
+	.regs = TWI1,
+	.irq_config = i2c1_sam_irq_config,
+	.periph_id = ID_TWI1,
+	.irq_id = TWI1_IRQn,
```
Comment (galak) [inline drivers/i2c/i2c_sam_twi.c]: use DTS for IRQ info

**Review by galak — CHANGES_REQUESTED** [review]
some minor dts cleanups, and avoid use of bitfield for i2c cfg stuff

File: drivers/i2c/i2c_sam_twi.c (hunk 2 above)
Comment (mnkp) [reply drivers/i2c/i2c_sam_twi.c]: As I understood there is a plan to move to generate configuration data structures directly from DTS thus the requirement. That sounds reasonable. On the other hand the IRQ, peripheral ID, register addresses data are all available already from header files for all CMSIS compliant arm SoCs. So now we need to re-write this data in DTS format. That's extra work and it's error prone. Maybe there is another way to handle it? Is the config structure generation concept finalized and ready?

**Comment by mnkp** [comment]
I have change usage of cfg.bits.speed as requested, also updated SAM4S Xplained board documentation adding support for I2C driver.

Could we handle the IRQ, peripheral ID, register addresses data from DTS issue when the config structure generation concept is ready? There are potentially more device drivers which will be affected.

**Review by dbkinder — DISMISSED** [review]
+1 for doc changes

**Comment by mnkp** [comment]
@galak I would like to rework twihs_sam I2C driver but it would create merge conflicts with this one if submitted simultaneously. Could we merge this driver if there are no further issues?

**Comment by galak** [comment]
I would like to make the DT changes so its easier to tell what needs to get generated and see what patterns there are as we go down that path.

**Comment by mnkp** [comment]
@galak the thing about IRQ number and register addresses is that we get them all for free from CMSIS header files. Going over DTS is a long way and is error prone; best avoided. However, if we want to generate device configuration data structure fully automatically then indeed DTS seems like the only way to go. The tricky part will be generating pin configuration data. I was asking about the concept because it's not clear to me what is the goal / what we want to achieve.

If we want to auto generate complete device configuration data structure then we also need to include Peripheral ID in DTS. Should I add it?

**Comment by galak** [comment]
Yes lets add Peripheral ID to the dts.  The intent is to get DTS to a point that SoC related data (address, irq, pins, clocks, etc) is coming from it, and we can use it to generate.  Which is why I ask to use the DT for the things we have, so we can start flushing out the things missing.

**Comment by bryan-hunt** [comment]
I've yet to see an example for the SAM series where the peripheral id differs from the IRQ number since the peripheral id is used to enable the interrupt. Might I suggest just using the IRQ number unless there is a compelling reason (like an exception to the rule) to add it?

**Comment by galak** [comment]
I think its better to introduce the peripheral id separate from the IRQ.  It saves us pain down the line when some SoC breaks the rule or have might have some peripheral with multiple IRQs.

**Comment by mnkp** [comment]
I've prepared requested update. I also added Peripheral ID to the configurations of all the remaining Atmel SAM DTS drivers (as a separate commit).

File: dts/arm/atmel/sam3x.dtsi [hunk 8]
```diff
@@ -49,14 +51,16 @@
...
 			status = "disabled";
 			label = "USART_1";
 		};
 
 		usart2: usart@400A0000 {
 			compatible = "atmel,sam-usart";
 			reg = <0x400A0000 0x130>;
-			interrupts = <18 0>;
+			interrupts = <19 0>;
+			pid = <19>;
```
Comment (galak) [inline dts/arm/atmel/sam3x.dtsi]: do you mind either putting the fix for uart in a separate commit, or at least comment in the commit message about fixing this.

File: dts/arm/atmel/sam3x.dtsi [hunk 9]
```diff
@@ -33,6 +33,7 @@
 			compatible = "atmel,sam-uart";
 			reg = <0x400E0800 0x124>;
 			interrupts = <8 0>;
+			pid = <8>;
```
Comment (galak) [inline dts/arm/atmel/sam3x.dtsi]: can we use peripheral-id instead of pid?

**Review by galak — COMMENTED** [review]
1. how about using 'peripheral-id'.  (pid is too generic and can mean a number of things).
2. minor nit about fix of uart irq, and commenting it or splitting the fix

File: dts/arm/atmel/sam3x.dtsi (hunk 8 above)
Comment (mnkp) [reply dts/arm/atmel/sam3x.dtsi]: Sorry about that, I put it now in a separate commit. s/pid/peripheral-id/g is done too.

File: boards/arm/arduino_due/Kconfig.defconfig [hunk 10]
```diff
@@ -40,20 +40,15 @@ endif # GPIO
 
 if I2C
 
-config I2C_ATMEL_SAM3
+config I2C_SAM_TWI
```
Comment (MaureenHelm) [inline boards/arm/arduino_due/Kconfig.defconfig]: Please add a comment in the commit message to note that you're changing the default I2C driver for this board

File: drivers/i2c/i2c_sam_twi.c [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+	Twi *const twi = dev_cfg->regs;
+	struct twi_msg *msg = &dev_data->msg;
+	u32_t isr_status;
+
+	/* Retrieve interrupt status */
+	isr_status = twi->TWI_SR & twi->TWI_IMR;
+
+	/* Not Acknowledged */
+	if (isr_status & TWI_SR_NACK) {
+		msg->twi_sr = isr_status;
```
Comment (MaureenHelm) [inline drivers/i2c/i2c_sam_twi.c]: Do you need to give the semaphore here?

Comment (mnkp) [reply drivers/i2c/i2c_sam_twi.c]: According to the datasheet TXCOMP is always set alongside NACK but the code indeed could have been written better. I updated it to make it more explicit.

File: boards/arm/arduino_due/Kconfig.defconfig (hunk 10 above)
Comment (mnkp) [reply boards/arm/arduino_due/Kconfig.defconfig]: Done.

**Review by galak — APPROVED** [review]

**Pull request merged**